src/vmspawn/vmspawn-util.c

```diff
@@ -0,0 +1,195 @@
+#include "architecture.h"
+#include "fd-util.h"
+#include "log.h"
+#include "macro-fundamental.h"
+#include "path-util.h"
+#include "strv.h"
+#include "vmspawn-util.h"
+#include <stdio.h>
+
+bool qemu_check_kvm_support(void) {
+        _cleanup_close_ int fd = -EBADF;
+        fd = open("/dev/kvm", O_RDONLY | O_CLOEXEC);
+        if (fd >= 0)
+                return true;
+
+        switch (errno) {
+        case ENOENT:
+                log_warning_errno(errno, "/dev/kvm not found. Not using KVM acceleration.");
+                break;
+
+        case EPERM:
+                _fallthrough_;
+        case EACCES:
+                log_warning_errno(errno, "Permission denied to access /dev/kvm. Not using KVM acceleration.");
+                break;
+        }
+
+        return false;
+}
+
+int find_ovmf_firmware(const char **ret_firmware_path) {
+#ifdef __x86_64__
+        #define FIRMWARE_LOCATIONS "/usr/share/ovmf/x64/OVMF_CODE.secboot.fd"
+#elif __i386__
+        #define FIRMWARE_LOCATIONS "/usr/share/edk2/ovmf-ia32/OVMF_CODE.secboot.fd", "/usr/share/OVMF/OVMF32_CODE_4M.secboot.fd"
+#else
+        #define FIRMWARE_LOCATIONS
+#endif
+
+        FOREACH_STRING(firmware, FIRMWARE_LOCATIONS) {
+                _cleanup_close_ int fd = -EBADF;
+                fd = open(firmware, O_CLOEXEC | O_RDONLY);
+                if (fd >= 0) {
+                        if (ret_firmware_path)
+                                *ret_firmware_path = firmware;
+                        return 0;
+                }
+        }
+
+#undef FIRMWARE_LOCATIONS
+#ifdef __x86_64__
+        #define FIRMWARE_LOCATIONS "/usr/share/ovmf/ovmf_code_x64.bin", "/usr/share/ovmf/x64/OVMF_CODE.fd", "/usr/share/qemu/ovmf-x86_64.bin"
+#elif __i386__
+        #define FIRMWARE_LOCATIONS "/usr/share/ovmf/ovmf_code_ia32.bin", "/usr/share/edk2/ovmf-ia32/OVMF_CODE.fd"
+#elif __aarch64__
+        #define FIRMWARE_LOCATIONS "/usr/share/AAVMF/AAVMF_CODE.fd"
+#elif __arm__
+        #define FIRMWARE_LOCATIONS "/usr/share/AAVMF/AAVMF32_CODE.fd"
+#else
+        #define FIRMWARE_LOCATIONS
+#endif
+
+        FOREACH_STRING(firmware, FIRMWARE_LOCATIONS) {
+                _cleanup_close_ int fd = -EBADF;
+                fd = open(firmware, O_CLOEXEC | O_RDONLY);
+                if (fd >= 0) {
+                        log_warning("Couldn't find OVMF firmware blob with secure boot support, "
+                                    "falling back to OVMF firmware blobs without secure boot support.");
+
+                        if (ret_firmware_path)
+                                *ret_firmware_path = firmware;
+                        return 1;
+                }
+        }
```

Comment:
i think this should be specified at build time of systemd and not auto-probed during runtime

the only reason mkosi probes for this, is because in python there's not really a "build time" per se...

but given that this is fixed per distro we should just expect downstreams to specify the right paths for this during build time as setting for meson

Reply:
btw, this should use access(F_OK) really, not open() to detect if some file exists

Reply:
On the first point I think neither my code not mkosi gets this correct. qemu defines that firmware should be searched for based on JSON configuration files following the format here: https://github.com/qemu/qemu/blob/master/docs/interop/firmware.json
The search path for these files include $XDG_CONFIG_HOME/qemu/firmware indicating that it should allow users to configure a custom firmware that qemu should use.
I think it would be best to adapt mkosi and vmspawn to support this interop format over statically specifying a specific firmware per distro?

Reply:
For now I have changed to using access and moved the logging messages out and into vmspawn.c
These fixes are included in fc2476205e.

Reply:
I'm going to keep working the JSON and search directories approach because I believe it to be the correct approach, if you disagree I will add it as a setting to meson instead.

Reply:
I have now implemented this.